File: secure-code-lab/examples/6_encapsulation/6.03_CWE-495_private-array-return/Vulnerable.java
/**
 * [취약 예제] Public 메소드로부터 반환된 Private 배열 (CWE-495 / KISA 6.03 캡슐화)
 *
 * 개념:
 *   객체 내부 상태(private 배열)를 public 메소드가 "참조 그대로" 반환하면,
 *   호출자는 캡슐화를 우회해 내부 배열을 직접 수정할 수 있다.
 *   자바 배열은 객체 참조이므로, 반환값을 바꾸면 원본 필드도 함께 바뀐다.
 *
 * 위험 지점:
 *   getScores() 가 this.scores 참조를 그대로 돌려준다.
 *   → 외부에서 반환 배열을 수정하면 내부 불변식이 깨진다.
 */
public class Vulnerable {

    // 내부 상태: 시험 점수 배열 (외부에서 직접 손대면 안 되는 값)
    private int[] scores;

    public Vulnerable(int[] initial) {
        // (별개 약점이지만) 여기서도 방어복사를 하는 것이 이상적이다.
        this.scores = initial;
    }

    /**
     * ★ 취약: private 배열의 참조를 그대로 반환한다.
     *   호출자가 반환값을 수정하면 내부 scores 도 같이 바뀐다(캡슐화 붕괴).
     */
    public int[] getScores() {
        return this.scores;
    }

    public int total() {
        int sum = 0;
        for (int s : scores) sum += s;
        return sum;
    }

    public static void main(String[] args) {
        Vulnerable v = new Vulnerable(new int[]{90, 80, 70});
        System.out.println("합계(정상) = " + v.total()); // 240

        // 외부 코드가 반환 배열을 조작 → 내부 상태 오염
        int[] leaked = v.getScores();
        leaked[0] = 0;

        System.out.println("합계(조작후) = " + v.total()); // 150 — 내부값이 바뀜
    }
}

File: secure-code-lab/examples/5_code-error/5.04_CWE-457_uninitialized-variable/Vulnerable.c
#include <stdio.h>

/*
 * [취약 예제] 초기화되지 않은 변수 사용 (CWE-457 / KISA 5.04)
 *
 * total 을 선언만 하고 초기값을 주지 않은 채 덧셈에 사용한다.
 * 지역 변수는 자동으로 0 이 되지 않으며, 스택에 남아 있던 쓰레기 값을
 * 그대로 갖는다. 따라서 합계 결과가 실행할 때마다 달라지고,
 * 이 값으로 크기 계산이나 분기를 하면 예측 불가능한 오작동이 생긴다.
 *
 * 위험 지점:
 *   int total;   ← 초기화 없는 선언. 이후 total 을 누산에 사용
 */

int sum_prices(const int *prices, int n) {
    /* ★ 취약: total 을 초기화하지 않았다. 시작값이 쓰레기 값이다. */
    int total;
    int i = n;

    while (i-- > 0) {
        total += prices[i];   /* 쓰레기 값에서 누산이 시작됨 → 결과 오염 */
    }
    return total;
}

int main(void) {
    int prices[] = {100, 200, 300};
    printf("sum=%d\n", sum_prices(prices, 3));
    return 0;
}
